src/tools/wasm-emscripten-finalize.cpp

```diff
@@ -233,7 +240,13 @@ int main(int argc, const char* argv[]) {
     }
   }
 
-  generator.generateDynCallThunks();
+  if (wasi) {
+    // Export a standard wasi "_start" method.
+    generator.exportWasiStart();
+  } else {
+    // If not wasi, then JS is relevant, and we need dynCalls
+    generator.generateDynCallThunks();
```

Comment:
What harm would generateDynCallThunks() do?  I guess I mean if you binary has constructs that require generateDynCallThunks would it not be already broken under wasi by definition?

Reply:
The direct harm is that emitting them makes it not a valid wasi file - wasmer, wasmtime etc. error with "invalid import" on them.

If we only emitted them when they were needed, then yeah, we could avoid this - and if they were emitted, it would indicate the wasm was already broken from wasi's point of view. But we emit these eagerly, for all function pointer signatures, basically - we emit anything JS might use (and the optimizer might remove them later).

Reply:
sgtm